x-pack/platform/plugins/shared/onechat/server/services/tools/builtin/definitions/create_visualization/create_visualization.ts

```diff
@@ -0,0 +1,163 @@
+const mapSchema = parse(mapAttributesSchema.getSchema());
+
+const createVisualizationSchema = z.object({
+  query: z.string().describe('A natural language query describing the desired visualization.'),
+  existingConfig: z
+    .string()
+    .optional()
+    .describe('An existing visualization configuration to modify.'),
+  chartType: z
+    .enum([SupportedChartType.Metric, SupportedChartType.Map])
+    .optional()
+    .describe(
+      '(optional) The type of chart to create as indicated by the user. If not provided, the LLM will suggest the best chart type.'
+    ),
+  esql: z
+    .string()
+    .optional()
+    .describe(
+      '(optional) An ES|QL query. If not provided, tool with automatically generate the query.'
+    ),
+});
+
+export const createVisualizationTool = (): BuiltinToolDefinition<
+  typeof createVisualizationSchema
+> => {
+  return {
+    id: platformCoreTools.createVisualization,
+    type: ToolType.builtin,
+    description: `Create a visualization configuration based on a natural language description.
+
+This tool will:
+1. Determine the best chart type if not specified (from: ${Object.values(SupportedChartType).join(
+      ', '
+    )})
+2. Generate an ES|QL query if not provided
+3. Generate a valid visualization configuration`,
+    schema: createVisualizationSchema,
+    tags: [],
+    handler: async (
+      { query: nlQuery, chartType, esql, existingConfig },
+      { esClient, modelProvider, logger, dataViewsService }
+    ) => {
+      try {
+        // Step 1: Determine chart type if not provided
+        let selectedChartType: SupportedChartType = chartType || SupportedChartType.Metric;
+        const parsedExistingConfig = existingConfig ? JSON.parse(existingConfig) : null;
+
+        if (!chartType) {
+          logger.debug('Chart type not provided, using LLM to suggest one');
+          const suggestedChartType = await getChartType(
+            modelProvider,
+            parsedExistingConfig?.type,
+            nlQuery
+          );
+          if (
+            Object.values(SupportedChartType).includes(suggestedChartType as SupportedChartType)
+          ) {
+            selectedChartType = suggestedChartType as SupportedChartType;
+          } else {
+            logger.warn(
+              `LLM suggested invalid chart type: ${suggestedChartType}, defaulting to metric`
+            );
+            selectedChartType = SupportedChartType.Metric;
+          }
+        }
+
+        // Step 2: Generate visualization configuration using langgraph with validation retry
+        const model = await modelProvider.getDefaultModel();
+
+        // Choose schema and graph based on chart type
+        const isMap = selectedChartType === SupportedChartType.Map;
+        const schema = isMap ? mapSchema : metricSchema;
+        const graph = isMap
+          ? createMapGraph(model, logger, esClient, dataViewsService)
+          : createVisualizationGraph(model, logger, esClient);
```

Comment:
So the way I was thinking about it, we would have one single graph, where we would handle all the visualization generation logic. Figuring out which chart type to use if unspecified (the call to `getChartType`) would be part of the graph, and then we would just fork between lens and map (and future other types) as sub graphs or subparts of that single graph.

I feel like it would have been better in terms of design (one single graph state, re-usability seems better that way and such).

Now I don't think I see anything blocker with the approach you implemented, so I'm also fine starting that way and figure out when we add new vis types / features to that flow if we think we should revisit.

Reply:
maps and lens visualizations are very different, thats why i split this out as we will probably want to update/improve the flows independently. lens graph is going to be responsible for generating metrics, gauges, xy charts, ..... so that one is already going to be complex enough imo.